File: 36.有效的数独.py
#
# @lc app=leetcode.cn id=36 lang=python3
#
# [36] 有效的数独
#

# @lc code=start
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        list_row = [{str(key): False for key in range(1, 10)} for _ in range(9)]
        list_col = [{str(key): False for key in range(1, 10)} for _ in range(9)]
        list_square = [{str(key): False for key in range(1, 10)} for _ in range(9)]

        for i in range(9):
            for j in range(9):
                number = board[i][j]
                if number == '.':
                    continue
                if list_row[i][number]:
                    return False
                else:
                   list_row[i][number] = True
                if list_col[j][number]:
                    return False
                else:
                    list_col[j][number] = True

                # if 0 <= i <= 2:
                #     if 0 <= j <= 2:
                #         k = 0
                #     elif 3 <= j <= 5:
                #         k = 1
                #     else:
                #         k = 2
                # elif  3 <= i <= 5:
                #     if 0 <= j <= 2:
                #         k = 3
                #     elif 3 <= j <= 5:
                #         k = 4
                #     else:
                #         k = 5
                # else:
                #     if 0 <= j <= 2:
                #         k = 6
                #     elif 3 <= j <= 5:
                #         k = 7
                #     else:
                #         k = 8
                
                k = (i // 3) * 3 + j // 3

                if list_square[k][number]:
                    return False
                else:
                    list_square[k][number] = True

        return True


        
# @lc code=end

File: 36.有效的数独.java
import java.util.Set;

/*
 * @lc app=leetcode.cn id=36 lang=java
 *
 * [36] 有效的数独
 */

// @lc code=start
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[][] rows = new int[9][9];
        int[][] cols = new int[9][9];
        int[][] sbox = new int[9][9];

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') continue;
                int n = board[i][j] - '1';
                int idx_box = (i/3)*3 + j/3;
                if (rows[i][n] == 1) return false;
                else rows[i][n] = 1;
                if (cols[j][n] == 1) return false;
                else cols[j][n] = 1;
                if (sbox[idx_box][n] == 1) return false;
                else sbox[idx_box][n] = 1;
            }
        }
        return true;

        // // 检查行
        //     for (int i = 0; i < 9; i++) {
        //         Set<Character> set = new HashSet<>();
        //         for (int j = 0; j < 9; j++) {
        //             if (board[i][j] == '.') continue;
        //             if (set.contains(board[i][j])) return false;
        //             else set.add(board[i][j]);
        //         }
        //     }
        // // 检查列
        //     for (int i = 0; i < 9; i++) {
        //         Set<Character> set = new HashSet<>();
        //         for (int j = 0; j < 9; j++) {
        //             if (board[j][i] == '.') continue;
        //             if (set.contains(board[j][i])) return false;
        //             else set.add(board[j][i]);
        //         }
        //     }
        // // 检查 3X3
        //     for (int i = 0; i < 9; i += 3) {
        //         for (int j = 0; j < 9; j += 3) {
        //             Set<Character> set = new HashSet<>();
        //             for (int p = 0; p < 3; p++) {
        //                 for (int q = 0; q < 3; q++) {
        //                     if (board[i+p][j+q] == '.') continue;
        //                     if (set.contains(board[i+p][j+q])) return false;
        //                     else set.add(board[i+p][j+q]);
        //                 }
        //             }
        //         }
        //     }


        // return true;
    }
}
// @lc code=end

